frame/assets/src/functions.rs

```diff
@@ -598,16 +619,13 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				Ok(())
 			})?;
 
+			source_died = Self::remove_if_dead(id, &source, &source_account, details);
+
 			// Remove source account if it's now dead.
-			if source_account.balance < details.min_balance {
-				debug_assert!(source_account.balance.is_zero(), "checked in prep; qed");
-				source_died =
-					Some(Self::dead_account(&source, details, &source_account.reason, false));
-				if let Some(Remove) = source_died {
-					Account::<T, I>::remove(id, &source);
-					return Ok(())
```

Comment:
I actually don't see why we need to extract this into a new function. In fact, most of this code here looks problematic, and can be simplified to:

```rust
if source_account.balance < details.min_balance {
    debug_assert!(source_account.balance.is_zero(), "checked in prep; qed");
    Self::dead_account(&source, details, &source_account.reason, false);
    Account::<T, I>::remove(id, &source);
    return Ok(())
}
```

The `source_died` binding can also be removed.